### PR TITLE
Wire structured-output row spans into paged path

### PR DESCRIPTION
**Pull request by CAICAIIs**
## Summary

This PR wires row-span-aware structured-output bitmasking into the Metal paged sampling path.

It builds on the paged decode segment metadata and keeps the row-span mapping, constrained-row construction, and bitmask application owned by `MetalStructuredOutputApplier`.

## What Changed

- Pass paged `decode_segments` from `MetalModelRunner._sample_paged_batch()` into `MetalStructuredOutputApplier.apply_paged()`.
- Map structured-output requests to one or more logits rows when paged speculative decode row spans are present.
- Apply per-position grammar bitmask rows for same-request structured output + scheduled speculative decode.
- Preserve the existing non-spec structured-output behavior.
- Add production/class-owned path coverage for:
  - normal decode structured output
  - prefill tail-row masking
  - mixed structured/plain speculative batches
  - request-id isolation
  - same-request structured output + scheduled speculative decode
  - missing per-position row-span bitmask rows

## Scope

This PR does not add a drafter, rejection sampling, KV rewind, or broader speculative decoding policy. It only wires structured-output row-span bitmasking into the existing Metal paged target path.

## Validation

- `pytest tests/test_grammar_bitmask.py tests/test_v1_model_runner_generate.py tests/test_spec_decode_metadata.py -q`
  - `66 passed`
- `ruff check` / `ruff format --check` on touched files
- `git diff --check`
- Real Metal smoke with `vllm 0.20.2`: `Qwen/Qwen3-0.6B` + paged attention + structured JSON output returned parseable JSON: `{"answer": 1}`

### CODE REVIEW

File: vllm_metal/v1/structured_output.py [hunk 1]
```diff
@@ -22,6 +24,187 @@
...
+
+    # rows_torch is CPU float32 (from torch.from_numpy), so torch_to_mlx goes
+    # through numpy — all xgrammar mutations are captured before the copy.
+    rows_mlx = torch_to_mlx(rows_torch).astype(original_dtype)  # (n_constrained, vocab)
+
+    # logits[0] produces a new lazy computation node (not a Python alias of
+    # logits), so __setitem__ here does not mutate the caller-held logits array.
+    result_2d = logits[0]
+    result_2d[logit_rows] = rows_mlx
+    return result_2d[None]  # Restore (1, total_tokens, vocab) shape
```
Comment (LxYuan0420) [inline vllm_metal/v1/structured_output.py]: This moves the core `MetalStructuredOutputApplier` behavior out into module-private helpers. These are not general structured-output utilities; they are the applier’s row-target mapping and bitmask-application internals. Please keep this behavior owned by `MetalStructuredOutputApplier` instead of spreading it across loose private functions at module scope.

Suggested shape:

```python
class MetalStructuredOutputApplier:
    def apply_paged(...): ...

    def _build_paged_row_targets(...): ...
    def _build_constrained_rows(...): ...
    def _apply_grammar_bitmask_to_rows(...): ...
```

File: vllm_metal/v1/structured_output.py [hunk 2]
```diff
@@ -104,67 +288,20 @@ def apply_paged(
...
-                constrained.append((req_id_to_row[req_id], bitmask_row))
-
-        if not constrained:
-            return logits
+        # Each tuple is (logit_row, bitmask_row).
+        constrained = _build_constrained_rows(
+            grammar_output.structured_output_request_ids,
+            row_targets,
+            grammar_bitmask_row_count=grammar_bitmask.shape[0],
+        )
```
Comment (LxYuan0420) [inline vllm_metal/v1/structured_output.py:123]: This introduces a row-span contract that the production `apply_paged()` path cannot consume yet. `model_runner` still builds one decode row per request, and this method still rejects same-request structured output plus speculative decoding before calling the mapping path. Since this PR is explicitly preparation only, I would rather not land dormant runtime logic. Please wire it through the real class-owned path in the PR that uses it, or keep it out until then.

File: vllm_metal/v1/structured_output.py [hunk 3]
```diff
@@ -22,6 +24,187 @@
...
+                constrained.append((row, bitmask_row + row_offset))
+        bitmask_row = end_bitmask_row
+
+    if bitmask_row != grammar_bitmask_row_count:
+        raise ValueError(
+            "Grammar bitmask row count must match row-span logits targets: "
+            f"consumed {bitmask_row}, got {grammar_bitmask_row_count}."
+        )
+
+    return constrained
```
Comment (LxYuan0420) [inline vllm_metal/v1/structured_output.py]: This row-span bitmask consumption logic should stay behind the same owner boundary as `apply_paged()`. Upstream vLLM already defines the `GrammarOutput` ordering and speculative bitmask expansion contract; Metal should adapt that contract in one class-owned place, not through standalone private functions that tests call directly.

File: tests/test_grammar_bitmask.py [hunk 4]
```diff
@@ -140,6 +143,36 @@ def _build_cu_seqlens(num_decode: int, prefill_lens: list[int]) -> list[int]:
...
+        grammar.structured_output_request_ids,
+        row_targets,
+        grammar_bitmask_row_count=grammar.grammar_bitmask.shape[0],
+        paged_decode_segments=segments,
+    )
+    return so._apply_grammar_bitmask_to_rows(
+        logits,
+        grammar.grammar_bitmask,
+        constrained,
+    )
```
Comment (LxYuan0420) [inline tests/test_grammar_bitmask.py]: These tests bypass the production owner by calling `so._build_paged_row_targets()`, `so._build_constrained_rows()`, and `so._apply_grammar_bitmask_to_rows()` directly. That proves helper mechanics, not the `MetalStructuredOutputApplier` contract. Please exercise a class-owned method instead.

File: tests/test_grammar_bitmask.py [hunk 5]
```diff
@@ -248,6 +281,126 @@ def test_mixed_decode_and_prefill(self) -> None:
...
+        with pytest.raises(NotImplementedError, match="one grammar bitmask row"):
+            _apply_row_span_grammar(
+                grammar,
+                decode_reqs,
+                [],
+                cu,
+                1,
+                logits,
+                segments,
+            )
```
Comment (LxYuan0420) [inline tests/test_grammar_bitmask.py:458]: These row-span tests validate the prepared private-helper path, but production `apply_paged()` still rejects same-request structured output plus speculative decoding. I do not want prep-only tests that pass by bypassing the runtime boundary. This coverage should land with the runtime PR that actually makes the row-span path reachable through the production owner.

**Review by LxYuan0420 — COMMENTED** [review]
Thanks for splitting this down. The direction makes sense, but I want this to land as a real production slice, not prep-only helpers. Please keep the current SO path unchanged here, and move the row-span work into the PR that wires SO + speculative decoding through `MetalStructuredOutputApplier`. That keeps ownership clear and tests on the production path.

**Comment by CAICAIIs** [comment]
> Thanks for splitting this down. The direction makes sense, but I want this to land as a real production slice, not prep-only helpers. Please keep the current SO path unchanged here, and move the row-span work into the PR that wires SO + speculative decoding through `MetalStructuredOutputApplier`. That keeps ownership clear and tests on the production path.

Thanks for the review! I understand the concern now. The current prep-only helper slice is not the right shape: the production path is still unreachable, and the tests bypass the MetalStructuredOutputApplier ownership boundary.

I’ll rework #358 into a real runtime slice: keep the current non-spec structured-output path unchanged, move the row-span mapping / constrained-row construction / bitmask application back under MetalStructuredOutputApplier as class-owned private methods, and make the tests exercise the class-owned or production path. I’ll also avoid duplicating #315’s DraftModelProposer / SpecDecodeWorker / rejection sampling / KV rewind scope; if #315 provides the final SD runtime contract, this PR will build the SO row-span bitmasking on top of that.

**Comment by ricky-chaoju** [comment]
Please resolve the conflict.

**Comment by CAICAIIs** [comment]
Thanks for the review. I resolved the conflict and reworked this PR from a prep-only helper slice into a reachable production slice.

Main changes:
- `MetalModelRunner._sample_paged_batch()` now passes paged `decode_segments` into `MetalStructuredOutputApplier.apply_paged()`.
- The row-span request-to-logits-row mapping, constrained-row construction, and bitmask application are now owned by `MetalStructuredOutputApplier`.
- Same-request structured output + scheduled speculative decode is now covered through the class-owned / production sampling path instead of standalone helper tests that bypass the runtime boundary.
- The existing non-spec structured-output behavior is preserved.

Validation:
- `pytest tests/test_grammar_bitmask.py tests/test_v1_model_runner_generate.py tests/test_spec_decode_metadata.py -q` -> 66 passed
- `ruff check` / `ruff format --check` on touched files -> passed
- `git diff --check` -> passed
- Real Metal smoke with `vllm 0.20.2`: `Qwen/Qwen3-0.6B` + paged attention + structured JSON output returned parseable JSON: `{"answer": 1}`

**Review by LxYuan0420 — APPROVED** [review]
Great work on this. I pushed a few small maintainer edits on top:

- Fail fast when row-span grammar output references a request with no paged logits rows, with a regression test.
- Document the row-span bitmask ordering contract from vLLM.
- Remove the redundant private `xgrammar` guard; `apply_paged()` owns that dependency check.

**Comment by CAICAIIs** [comment]
> Great work on this. I pushed a few small maintainer edits on top:
> 
> * Fail fast when row-span grammar output references a request with no paged logits rows, with a regression test.
> * Document the row-span bitmask ordering contract from vLLM.
> * Remove the redundant private `xgrammar` guard; `apply_paged()` owns that dependency check.

Thanks!!
